src/id.rs

```diff
@@ -279,51 +280,46 @@ impl Display for P2pNode {
     }
 }
 
-impl PartialOrd for P2pNode {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
+#[derive(Clone, Serialize, Deserialize)]
+struct OrderedConnectionInfo(pub ConnectionInfo);
+
+impl Hash for OrderedConnectionInfo {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let self_addr = deconstruct_socket_addr(&self.0.peer_addr);
+        self_addr.hash(state);
+        self.0.peer_cert_der.hash(state);
     }
 }
 
-impl Ord for P2pNode {
-    fn cmp(&self, other: &Self) -> Ordering {
-        let self_socket_addr = SocketAddrWrapper(self.connection_info.peer_addr);
-        let other_socket_addr = SocketAddrWrapper(other.connection_info.peer_addr);
-        (
-            &self.public_id,
-            &self_socket_addr,
-            &self.connection_info.peer_cert_der,
-        )
-            .cmp(&(
-                &other.public_id,
-                &other_socket_addr,
-                &other.connection_info.peer_cert_der,
-            ))
+impl PartialEq for OrderedConnectionInfo {
+    fn eq(&self, other: &Self) -> bool {
+        let self_addr = deconstruct_socket_addr(&self.0.peer_addr);
+        let other_addr = deconstruct_socket_addr(&other.0.peer_addr);
+        (self_addr, &self.0.peer_cert_der).eq(&(other_addr, &other.0.peer_cert_der))
     }
 }
 
-#[derive(PartialEq, Eq)]
-struct SocketAddrWrapper(pub SocketAddr);
+impl Eq for OrderedConnectionInfo {}
 
-impl PartialOrd for SocketAddrWrapper {
+impl PartialOrd for OrderedConnectionInfo {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
-impl Ord for SocketAddrWrapper {
+
+impl Ord for OrderedConnectionInfo {
     fn cmp(&self, other: &Self) -> Ordering {
-        compare_socket_addr(self.0, other.0)
+        let self_addr = deconstruct_socket_addr(&self.0.peer_addr);
+        let other_addr = deconstruct_socket_addr(&other.0.peer_addr);
+        (self_addr, &self.0.peer_cert_der).cmp(&(other_addr, &other.0.peer_cert_der))
     }
 }
 
-fn compare_socket_addr(left_addr: SocketAddr, right_addr: SocketAddr) -> Ordering {
-    let to_tuple = |socket_addr: SocketAddr| match socket_addr {
+fn deconstruct_socket_addr(socket_addr: &SocketAddr) -> (Ipv6Addr, u16, u32, u32) {
```

Comment:
If you include the &peer_cert_der to the return result, it will be much simpler to implement the Ord/Eq/Hash.

Reply:
Agree, done!